src/components/data/QueryAuth0.js

```diff
@@ -37,24 +38,82 @@ export default class QueryAuth0 extends React.PureComponent {
+        // accessToken refresh loop can silently authenticate you again.
+        startAccessTokenRefreshLoop = true;
+      }
+      // But this is only good enough if it hasn't expired.
+      const expiresAt = JSON.parse(localStorage.getItem('expiresAt'));
+      if (expiresAt) {
+        if (expiresAt - new Date().getTime() < 0) {
+          // Oh no! It has expired.
+          authResult = null;
+        }
+      }
     }
 
     if (authResult) {
-      const { state } = authResult;
-      logUserIn(authResult);
-      // Since we include 'id_token' for the 'responseType' in auth0.WebAuth
-      // the authresult will contain the user profile as .idTokenPayload
-      // included with the accessToken. Use this now to update the state so that we
-      // can display the name and avatar you logged in as.
-      userProfileReceived(authResult.idTokenPayload);
+      // This is only true if you arrived on the site with a valid window.location.hash
+      // or the localStorage.authResult (and localStorage.expiresAt) wasn't out of date.
+      this.postProcessAuthResult(authResult);
+    }
+    if (startAccessTokenRefreshLoop) {
+      // Start a never-ending loop of periodically checking if the accessToken is about to
+      // or has expired.
+      // This only happens for users who have successfully logged in or have successfully
+      // logged in the *past* but expired.
+      await this.accessTokenRefreshLoop();
+    }
+  }
+
+  postProcessAuthResult = authResult => {
+    const { logUserIn, userProfileReceived } = this.props;
+    const { state } = authResult;
+    logUserIn(authResult);
+    // Since we include 'id_token' for the 'responseType' in auth0.WebAuth
+    // the authresult will contain the user profile as .idTokenPayload
+    // included with the accessToken. Use this now to update the state so that we
+    // can display the name and avatar you logged in as.
+    userProfileReceived(authResult.idTokenPayload);
+
+    if (state) {
+      this.props.push(state);
+    }
+  };
+
+  async accessTokenRefreshLoop() {
+    if (!localStorage.getItem('expiresAt')) {
+      // The user has logged out probably. Either way, no point bothering to refresh.
+      return;
+    }
+    const expiresAt = JSON.parse(localStorage.getItem('expiresAt'));
+    const left = expiresAt - new Date().getTime();
+    const preemptive = REFRESH_AUTH_PREEMPTIVE_SECONDS * 1000;
 
-      if (state) {
-        this.props.push(state);
+    const accessTokenRefreshLoopTimer = window.setTimeout(async () => {
+      await this.accessTokenRefreshLoop();
+    }, REFRESH_AUTH_PERIOD_SECONDS * 1000);
+    if (left - preemptive < 0) {
+      // Time to refresh!
+      console.warn('Time to refresh auth session');
+      try {
+        const authResult = await refreshAuthentication(window.location.pathname);
+        this.postProcessAuthResult(authResult);
+      } catch (err) {
+        window.clearTimeout(accessTokenRefreshLoopTimer);
+        console.error(err);
```

Comment:
I don't think we need to put this in here since we're handling this error on the next line.